### PR TITLE
Default to updated version sorting

### PR DESCRIPTION
**Pull request by ljacomet**

### CODE REVIEW

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test QF

**Comment by ljacomet** [comment]
@bot-gradle test QF

**Comment by ljacomet** [comment]
@bot-gradle test QF

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java [hunk 1]
```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
...
             }
-            if (releaseBias) {
+            // Only care about the first qualified version that matches
+            if (bestComponent == null) {
                 ComponentResolveMetadata metaData = component.getMetadata();
                 if (metaData != null && "release".equals(metaData.getStatus())) {
                     details.select(component);
                     return;
                 }
+                // Special case Maven snapshots
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java:83]: This became more complicated than intended, because of unexpected impact on versions. So instead of the "general" solution to get the right behavior for `SNAPSHOT`, this is changed to be _specific_ for `SNAPSHOT`.

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/strategy/DefaultVersionComparatorTest.groovy [hunk 2]
```diff
@@ -49,6 +292,8 @@ class DefaultVersionComparatorTest extends AbstractDefaultVersionComparatorTest
 
         where:
         smaller | larger
-        "1.0-release"   | "1.0-final"
+        "1.0-final"   | "1.0-release"
+        "1.0-final"   | "1.0-ga"
+        "1.0-ga"   | "1.0-release"
```
Comment (jjohannes) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/strategy/DefaultVersionComparatorTest.groovy:297]: Table formatting is off here

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java [hunk 3]
```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
...
+                if (metaData instanceof MavenModuleResolveMetadata) {
+                    if (((MavenModuleResolveMetadata) metaData).getSnapshotTimestamp() != null || metaData.getModuleVersionId().getVersion().endsWith("-SNAPSHOT")) {
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java:85]: Maybe put this test (combining both `if`s) into a `isMavenSnapshot()` method. Would make this quite complex piece of code easier to read.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java [hunk 4]
```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
...
+                    return;
+                } else if (unqalifiedComponent == null) {
+                    unqalifiedComponent = component;
+                }
             }
-            if (releaseBias) {
+            // Only care about the first qualified version that matches
+            if (bestComponent == null) {
                 ComponentResolveMetadata metaData = component.getMetadata();
                 if (metaData != null && "release".equals(metaData.getStatus())) {
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java:79]: Maybe put this also into a `hasReleaseStatus()` method for readability (see comment about the other "if" below)

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java [hunk 5]
```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
...
+                    }
+                }
             }
         }
 
-        // Nothing - just return the highest version
-        details.select(matches.get(sorted.get(0)));
+        if (bestComponent != null) {
+            if (unqalifiedComponent != null) {
+                details.select(unqalifiedComponent);
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java:94]: Wasn't the previous case that we select the "unqualifiedComponent" event if "bestComponent == null"? I guess we do not reach this situation because we return earlier already. In any case I would simplify this here (see next comment).

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java [hunk 6]
```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
         // Work backwards from highest version, return the first candidate with qualified version and release status, or candidate with unqualified version
         List<Version> sorted = new ArrayList<>(matches.keySet());
         sorted.sort(Collections.reverseOrder(versionComparator));
+        T bestComponent = null;
+        T unqalifiedComponent = null;
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java:65]:
```suggestion
        T unqualifiedComponent = null;
```

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java [hunk 7]
```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
...
+        if (bestComponent != null) {
+            if (unqalifiedComponent != null) {
+                details.select(unqalifiedComponent);
+                return;
+            }
+            details.select(bestComponent);
+        } else {
+            // Nothing - just return the highest version
+            details.select(matches.get(sorted.get(0)));
+        }
```
Comment (jjohannes) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java:101]: I think this would be easier to read like this:

```suggestion
        if (unqualifiedComponent != null) { // as by my earlier comment I think we do not need `&& bestComponent != null` here
            // Comment explaining this case....
            details.select(unqualifiedComponent);     
            return;
       }
       if (bestComponent != null) {
            // Comment explaining this case...
            details.select(bestComponent);
            return;
        }
        // Nothing - just return the highest version
        details.select(matches.get(sorted.get(0)));
      
```

**Review by jjohannes — CHANGES_REQUESTED** [review]
Good work. The algorithm looks right also it is hard to follow. Therefore I made some suggestion to improve the readability for our future selves. Maybe you can also add some more explanation in the code (see my comments for details).

**Comment by ljacomet** [comment]
@jjohannes Thanks for the review. I have addressed all the comments.

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Review by jjohannes — APPROVED** [review]
Thanks for the additional cleanup/documentation work @ljacomet. Looks great.

**Comment by ljacomet** [comment]
@bot-gradle test RFM
